hash-table/open-addressing.js

```diff
@@ -0,0 +1,129 @@
+const DEFAULT_CAPACITY = 8
+const UPPER_LOAD_FACTOR = 0.7
+const LOWER_LOAD_FACTOR = 0.25
+const GROW_SHRINK_FACTOR = 2
+
+class Node {
+  constructor(key, value, deleted = false) {
+    this.key = key
+    this.value = value
+    this.deleted = deleted
+  }
+}
+
+class OpenAddressingHT {
+  constructor() {
+    this.arr = new Array(DEFAULT_CAPACITY)
+    this.capacity = DEFAULT_CAPACITY
+    this.size = 0
+  }
+
+  insert(key, value) {
+    const { curNode, index } = this.__search(key)
+
+    if (!curNode || curNode.deleted) {
+      const node = new Node(key, value)
+      this.arr[index] = node
+      this.size++
+      if (this.__getGrowShrinkFactor() >= UPPER_LOAD_FACTOR) {
+        this.__grow()
+      }
+      return
+    } 
+
+    if (curNode.key === key) {
+      this.arr[index].value = value
+      return
+    }
+  }
+
+  search(key) {
+    const { curNode } = this.__search(key)
+    return curNode ? curNode.value : curNode
+  }
+
+  delete(key) {
+    const { curNode, index } = this.__search(key)
+    if (!curNode) return 
+  
+    this.arr[index].deleted = true
+    this.size--
+    if (this.__getGrowShrinkFactor() <= LOWER_LOAD_FACTOR) {
+      this.__shrink()
+    }
+  }
+
+  __search(key) {
+    let trial = 1
+
+    while (trial <= this.capacity) {
+      const index = this.__doubleHash(key, trial)
+      const curNode = this.arr[index]
+
+      if (!curNode) return { curNode: null, index }
+
+      if (curNode.key === key && !curNode.deleted) {
+        return { curNode, index }
+      }
+      trial++
+    }
+  }
+
+  __grow() {
+    const newCapacity = this.capacity * GROW_SHRINK_FACTOR
+    this.__rehash(newCapacity)
+  }
+
+  __shrink() {
+    const newCapacity = this.capacity / GROW_SHRINK_FACTOR
+    this.__rehash(newCapacity)
+  }
+
+  __rehash(newCapacity) {
+    const oldCapacity = this.capacity
+    const oldArr = [...this.arr]
```

Comment:
You don't really have to do this. The spread operator `...` here will loop over `this.arr` and copy its elements **by reference** to `oldArr`. It takes `O(n)`.

Instead, just assign `oldArr = this.arr`. Later down when you assign `this.arr = new Array()`, the `oldArr` still holds the reference to the old bucket array, and no `O(n)` operation takes place here.